src/server/rewindable.rs

```diff
@@ -0,0 +1,204 @@
+{
+    #[inline]
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        self.inner.prepare_uninitialized_buffer(buf)
+    }
+
+    #[inline]
+    fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
+        if let Some(bs) = self.pre.take() {
+            // If there are no remaining bytes, let the bytes get dropped.
+            if bs.len() > 0 {
+                let mut reader = bs.into_buf().reader();
+                let mut writer = buf.writer();
+                let cnt = io::copy(&mut reader, &mut writer)? as usize;
+
+                let mut new_pre = reader.into_inner().into_inner();
+                new_pre.advance(cnt);
+
+                // Put back whats left
+                if new_pre.len() > 0 {
+                    self.pre = Some(new_pre);
+                }
+
+                return Ok(Async::Ready(cnt));
+            }
+        }
+        self.inner.read_buf(buf)
+    }
+}
+
+impl<T> AsyncWrite for Rewindable<T>
+where
+    T: AsyncWrite,
+{
+    #[inline]
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        AsyncWrite::shutdown(&mut self.inner)
+    }
+
+    #[inline]
+    fn write_buf<B: Buf>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
+        self.inner.write_buf(buf)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    // use super::*;
+    extern crate tokio_mockstream;
+    use self::tokio_mockstream::MockStream;
+    use std::io::Cursor;
+
+    #[test]
+    fn async_partial_rewind() {
+        let bs = &mut [104, 101, 108, 108, 111];
+        let o1 = &mut [0, 0];
+        let o2 = &mut [0, 0, 0, 0, 0];
+
+        let mut stream = Rewindable::new(MockStream::new(bs));
+        let mut o1_cursor = Cursor::new(o1);
+        match stream.read_buf(&mut o1_cursor).unwrap() {
+            Async::NotReady => panic!("should be ready"),
+            Async::Ready(cnt) => assert_eq!(2, cnt),
+        }
+
+        let read_buf = Bytes::from(&o1_cursor.into_inner()[..]);
+        stream.prepend(read_buf);
+
+        let mut o2_cursor = Cursor::new(o2);
+        stream.read_buf(&mut o2_cursor).unwrap();
+        stream.read_buf(&mut o2_cursor).unwrap();
+        let o2_final = o2_cursor.into_inner();
+
+        assert_eq!(&o2_final, &bs);
+    }
+    #[test]
+    fn async_full_rewind() {
+        let bs = &mut [104, 101, 108, 108, 111];
+        let o1 = &mut [0, 0, 0, 0, 0];
+        let o2 = &mut [0, 0, 0, 0, 0];
+
+        let mut stream = Rewindable::new(MockStream::new(bs));
+        let mut o1_cursor = Cursor::new(o1);
+        match stream.read_buf(&mut o1_cursor).unwrap() {
+            Async::NotReady => panic!("should be ready"),
+            Async::Ready(cnt) => assert_eq!(5, cnt),
+        }
+
+        let read_buf = Bytes::from(&o1_cursor.into_inner()[..]);
+        stream.prepend(read_buf);
+
+        let mut o2_cursor = Cursor::new(o2);
+        stream.read_buf(&mut o2_cursor).unwrap();
+        stream.read_buf(&mut o2_cursor).unwrap();
```

Comment:
@seanmonstar This the crux of what im concerned about, is it okay that we're not reading out all bytes from a perf/correctness point of view?

Reply:
I don't actually understand the question, or what this test is testing :D

Mind adding some comments about what each part of the test is triggering or why?

Reply:
Haha sorry, was a bit of a non-question. I think ive answered it myself, I was wondering if there is an undocumented dependency of the `read_buf` method to read all the bytes possible into the `buf` in a single call.
How I'm doing it here is *one* of the following per call:
- Is there a prefix? Read as many bytes in the prefix as possible out, If there are some left, put those back in the prefix field.
- If there is not a prefix, read direct from the `inner`

In practice this means the buf passed in will not always be filled even though there may be bytes ready to read. I can imagine a situation where there is code relying on a small buffer being completely filled which could break.

Reply:
Oh, I see what you mean. Yea, this behavior is correct. You don't want to copy the buffered bytes, and try to a syscall in the same `read`. Because: what's the behavior if there is an error from the syscall? Return the error? But then the buffered bytes... what about those? It's messy.

Plus, anyone *relying* on a `read` to completely fill a buffer is definitely using IO incorrectly. You never know exactly how much data might be in the kernels buffers, and you don't know if more arrived just as it partially filled your buffer, so you just have to keep reading until your buffer is full, or your get `WouldBlock`.